apis/Google.Cloud.Firestore/Google.Cloud.Firestore/Converters/ValueTupleConverter.cs

```diff
@@ -0,0 +1,444 @@
+        }
+
+        /// <summary>
+        /// Creates a converter for the given property if it is a suitable property for a value tuple conversion.
+        /// </summary>
+        internal static ValueTupleConverter MaybeCreateConverter(PropertyInfo property)
+        {
+            var attribute = property.GetCustomAttribute<TupleElementNamesAttribute>(true);
+            if (attribute == null)
+            {
+                return null;
+            }
+            var names = attribute.TransformNames;
+            if (names.Any(string.IsNullOrEmpty))
+            {
+                return null;
+            }
+            var propertyType = property.PropertyType;
+            // If the property is actually nullable, we unwrap that and proceed with the non-nullable form.
+            // Everything just works simply.
+            var nonNullableType = Nullable.GetUnderlyingType(propertyType) ?? propertyType;
+            if (!nonNullableType.IsConstructedGenericType)
+            {
+                return null;
+            }
+            var genericTypeDefinition = nonNullableType.GetGenericTypeDefinition();
+            if (!s_genericTupleTypes.Contains(genericTypeDefinition))
+            {
+                return null;
+            }
+            return new ValueTupleConverter(nonNullableType, names);
+        }
+
+        public override object DeserializeMap(DeserializationContext context, IDictionary<string, Value> values)
+        {
+            // TODO: What if the keys in the map don't match our names, or there are spare/missing ones?
+            var accessor = CreateAccessor(Activator.CreateInstance(TargetType));
+            for (int i = 0; i < _names.Count; i++)
+            {
+                if (values.TryGetValue(_names[i], out var value))
+                {
+                    accessor[i] = ValueDeserializer.Deserialize(context, value, _elementTypes[i]);
+                }
+            }
+            return accessor.Value;
+        }
+
+        public override void SerializeMap(SerializationContext context, object value, IDictionary<string, Value> map)
+        {
+            var accessor = CreateAccessor(value);
+            for (int i = 0; i < _names.Count; i++)
+            {
+                map[_names[i]] = ValueSerializer.Serialize(context, accessor[i]);
+            }
+        }
+
+        private ITupleAccessor CreateAccessor(object obj)
+        {
+            ITupleAccessor accessor = (ITupleAccessor) Activator.CreateInstance(_accessorType);
+            accessor.Value = obj;
+            return accessor;
+        }
+
+        private interface ITupleAccessor
+        {
+            object this[int index] { get; set; }
+            object Value { get; set; }            
+        }
+
+        // Tuple accessor implementations. These all look the same, 
+        private sealed class TupleAccessor<T1> : ITupleAccessor
```

Comment:
Might using `System.Runtime.CompilerServices.ITuple` mean these `TupleAccessor<...>` classes are not required?

Reply:
It would indeed. Unfortunately it's not available in netstandard2.0, despite the docs :( (It's in 2.1.)

Reply:
Oh right, that's unfortunate (and wrongly documented!).